diff -uNr old/net/mac80211/rate.c new/net/mac80211/rate.c
--- old/net/mac80211/rate.c	2024-12-27 00:16:46.000000000 +0800
+++ new/net/mac80211/rate.c	2025-03-02 10:33:47.938561000 +0800
@@ -100,13 +100,19 @@
 	struct ieee80211_sta *ista = &sta->sta;
 	void *priv_sta = sta->rate_ctrl_priv;
 	struct ieee80211_chanctx_conf *chanctx_conf;
+	struct ieee80211_link_data *link;
 
 	WARN_ON(link_id != 0);
 
 	if (ref && ref->ops->rate_update) {
 		rcu_read_lock();
 
-		chanctx_conf = rcu_dereference(sta->sdata->vif.bss_conf.chanctx_conf);
+		link = rcu_dereference(sta->sdata->link[link->link_id]);
+		if (!link) {
+			rcu_read_unlock();
+			return;
+		}
+		chanctx_conf = rcu_dereference(link->conf->chanctx_conf);
 		if (WARN_ON(!chanctx_conf)) {
 			rcu_read_unlock();
 			return;
