diff -uNr old/net/mac80211/eht.c new/net/mac80211/eht.c
--- old/net/mac80211/eht.c	2024-12-27 00:16:47.000000000 +0800
+++ new/net/mac80211/eht.c	2025-03-02 10:41:02.516538000 +0800
@@ -75,4 +75,7 @@
 
 	link_sta->cur_max_bandwidth = ieee80211_sta_cap_rx_bw(link_sta);
 	link_sta->pub->bandwidth = ieee80211_sta_cur_vht_bw(link_sta);
+	/* 5G and 6G ignore the max mpdu length in eht cap */
+	if (sband->band != NL80211_BAND_2GHZ)
+		return;
 }
