diff -uNr old/include/linux/ieee80211.h new/include/linux/ieee80211.h
--- old/include/linux/ieee80211.h	2025-03-03 22:55:43.655563000 +0800
+++ new/include/linux/ieee80211.h	2025-03-03 23:02:01.960523000 +0800
@@ -3183,11 +3183,6 @@
 {
 	u8 count = 0;
 
-	/* on 2.4 GHz, if it supports 40 MHz, the result is 3 */
-	if (he_cap->phy_cap_info[0] &
-	    IEEE80211_HE_PHY_CAP0_CHANNEL_WIDTH_SET_40MHZ_IN_2G)
-		return 3;
-
 	/* on 2.4 GHz, these three bits are reserved, so should be 0 */
 	if (he_cap->phy_cap_info[0] &
 	    IEEE80211_HE_PHY_CAP0_CHANNEL_WIDTH_SET_40MHZ_80MHZ_IN_5G)
@@ -3202,6 +3197,10 @@
 
 	if (count)
 		return count;
+	/* on 2.4 GHz, if it supports 40 MHz, the result is 3 */
+	if (he_cap->phy_cap_info[0] &
+		IEEE80211_HE_PHY_CAP0_CHANNEL_WIDTH_SET_40MHZ_IN_2G)
+		return 3;
 
 	return from_ap ? 3 : 4;
 }
