diff -uNr old/net/mac80211/cfg.c new/net/mac80211/cfg.c
--- old/net/mac80211/cfg.c	2025-03-03 23:56:33.728428000 +0800
+++ new/net/mac80211/cfg.c	2025-03-03 23:59:48.748402000 +0800
@@ -1632,11 +1632,11 @@
 
 	ieee80211_free_next_beacon(link);
 
-	/* turn off carrier for this interface and dependent VLANs */
-	list_for_each_entry(vlan, &sdata->u.ap.vlans, u.vlan.list)
-		netif_carrier_off(vlan->dev);
 
 	if (ieee80211_num_beaconing_links(sdata) <= 1) {
+		/* turn off carrier for this interface and dependent VLANs */
+		list_for_each_entry(vlan, &sdata->u.ap.vlans, u.vlan.list)
+			netif_carrier_off(vlan->dev);		
 		netif_carrier_off(dev);
 		sdata->u.ap.active = false;
 	}
