diff -uNr old/net/wireless/mlme.c new/net/wireless/mlme.c
--- old/net/wireless/mlme.c	2025-03-03 23:56:32.894293000 +0800
+++ new/net/wireless/mlme.c	2025-03-04 00:03:28.331306000 +0800
@@ -1112,7 +1112,8 @@
 	struct cfg80211_registered_device *rdev = wiphy_to_rdev(wiphy);
 	unsigned long timeout;
 
-	if (WARN_ON(wdev->valid_links &&
+	if (WARN_ON(event != NL80211_RADAR_CAC_ABORTED &&	    
+		wdev->valid_links &&
 		    !(wdev->valid_links & BIT(link_id))))
 		return;
 
