diff -uNr old/net/mac80211/mlme.c new/net/mac80211/mlme.c
--- old/net/mac80211/mlme.c	2025-03-04 00:42:30.684119000 +0800
+++ new/net/mac80211/mlme.c	2025-03-04 00:49:23.558072000 +0800
@@ -2452,10 +2452,6 @@
 			}
 			break;
 		case IEEE80211_CSA_SOURCE_OTHER_LINK:
-			/* active link: we want to see the beacon to continue */
-			if (ieee80211_vif_link_active(&sdata->vif,
-						      link->link_id))
-				return;
 
 			/* switch work ran, so just complete the process */
 			if (link->u.mgd.csa.waiting_bcn) {
@@ -2468,6 +2464,10 @@
 				 */
 				break;
 			}
+			/* active link: we want to see the beacon to continue */
+			if (ieee80211_vif_link_active(&sdata->vif,
+							  link->link_id))
+				return;
 
 			/* link still has CSA but we already know, do nothing */
 			if (!res)
@@ -6687,7 +6687,7 @@
 		struct ieee80211_mle_per_sta_profile *prof;
 		struct ieee802_11_elems *prof_elems;
 		struct ieee80211_link_data *link;
-		ssize_t len;
+		ssize_t len, capab_info_len = sizeof(u16);
 
 		if (link_id == reporting_link_id)
 			continue;
@@ -6711,10 +6711,12 @@
 			continue;
 
 		prof = (void *)sta_profiles[link_id];
+		/* skip capab info (u16) */
 		prof_elems = ieee802_11_parse_elems(prof->variable +
-						    (prof->sta_info_len - 1),
-						    len -
-						    (prof->sta_info_len - 1),
+							(prof->sta_info_len - 1) +
+							capab_info_len,
+							len - (prof->sta_info_len - 1) -
+							capab_info_len,
 						    false, NULL);
 
 		/* memory allocation failed - let's hope that's transient */
