--- a/npu.c
+++ b/npu.c
@@ -497,6 +497,10 @@ int mt76_npu_init(struct mt76_dev *dev)
 	struct airoha_npu *npu;
 	int err = 0;
 
+	/* NPU offloading is only supported for MT7992 */
+	if (!is_mt7992(dev))
+		return -EINVAL;
+
 	mutex_lock(&dev->mutex);
 
 	npu = airoha_npu_get(dev->dev, NULL);
--- a/mt7996/pci.c
+++ b/mt7996/pci.c
@@ -257,6 +257,9 @@ static int mt7996_pci_probe(struct pci_d
 
 	//dev->has_rro = rro_enable;
 	mdev = &dev->mt76;
+	if (!is_mt7992(mdev))
+		rro_mode = MT76_HWRRO_DISABLE;
+
 	switch (rro_mode) {
 	case MT76_HWRRO_DISABLE:
 	case MT76_HWRRO_V3:
