--- a/npu.c
+++ b/npu.c
@@ -3,6 +3,7 @@
  * Copyright (c) 2025 AIROHA Inc
  * Author: Lorenzo Bianconi <lorenzo@kernel.org>
  */
+#include <linux/kernel.h>
 #include <net/flow_offload.h>
 #include <net/pkt_cls.h>
 
@@ -13,6 +14,9 @@
 #define MT76_NPU_RX_BUF_SIZE	(1800 + \
 				 SKB_DATA_ALIGN(sizeof(struct skb_shared_info)))
 
+static bool npu_enable = true;
+module_param(npu_enable, bool, 0644);
+
 static int mt76_npu_fill_rx_queue(struct mt76_dev *dev, struct mt76_queue *q)
 {
 	int nframes = 0;
@@ -504,6 +508,9 @@ int mt76_npu_init(struct mt76_dev *dev)
 	if (!is_mt7992(dev))
 		return -EINVAL;
 
+	if (!npu_enable)
+		return -EINVAL;
+
 	mutex_lock(&dev->mutex);
 
 	npu = airoha_npu_get(dev->dev, NULL);
