diff -uNr old/mt7996/mcu.c new/mt7996/mcu.c
--- old/mt7996/mcu.c	2025-05-31 13:32:06.046951000 +0800
+++ new/mt7996/mcu.c	2025-05-31 13:10:27.117026000 +0800
@@ -4680,11 +4680,10 @@
 		ret = -EINVAL;
 		goto out;
 	}
-
 	hdr = (const void *)(fw->data + fw->size - sizeof(*hdr));
 	memcpy(dev->ram_build_date[ram_type],
 	       hdr->build_date,
-	       sizeof(dev->ram_build_date[ram_type]));
+	       sizeof(dev->ram_build_date[ram_type])> sizeof(hdr->build_date) ? sizeof(hdr->build_date) : sizeof(dev->ram_build_date[ram_type]));
 	dev_info(dev->mt76.dev, "%s Firmware Version: %.10s, Build Time: %.15s\n",
 		 fw_type, hdr->fw_ver, hdr->build_date);
 
