diff -uNr old/mt76.h new/mt76.h
--- old/mt76.h	2025-06-23 15:25:15.275541000 +0800
+++ new/mt76.h	2025-06-23 15:04:05.470617000 +0800
@@ -377,6 +377,7 @@
 	MT_PHY_TYPE_HT,
 	MT_PHY_TYPE_HT_GF,
 	MT_PHY_TYPE_VHT,
+	MT_PHY_TYPE_PLR, /* Proprietary Long Range */
 	MT_PHY_TYPE_HE_SU = 8,
 	MT_PHY_TYPE_HE_EXT_SU,
 	MT_PHY_TYPE_HE_TB,
diff -uNr old/mt7996/main.c new/mt7996/main.c
--- old/mt7996/main.c	2025-06-23 15:25:15.117537000 +0800
+++ new/mt7996/main.c	2025-06-23 15:11:53.290585000 +0800
@@ -1927,24 +1927,31 @@
 	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
 	struct mt7996_link_sta *mlink;
-	struct rate_info *rate;
+	//struct rate_info *rate;
 
 	mutex_lock(&dev->mt76.mutex);
 	mlink = mlink_dereference_protected(msta, msta->pri_link);
 	if (!mlink)
 		goto out;
 
+#if 0
 	rate = &mlink->wcid.rate;
 	if (rate->legacy || rate->flags) {
 		sinfo->txrate = *rate;
 		sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_BITRATE);
 	}
-
+#endif
+	sinfo->txrate = mlink->wcid.rate;
+	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_BITRATE);
+#if 0	
 	rate = &mlink->wcid.rx_rate;
 	if (rate->legacy || rate->flags) {
 		sinfo->rxrate = *rate;
 		sinfo->filled |= BIT_ULL(NL80211_STA_INFO_RX_BITRATE);
 	}
+#endif
+	sinfo->rxrate = mlink->wcid.rx_rate;
+	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_RX_BITRATE);
 
 	sinfo->tx_failed = mlink->wcid.stats.tx_failed;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_FAILED);
diff -uNr old/mt7996/mcu.c new/mt7996/mcu.c
--- old/mt7996/mcu.c	2025-06-23 15:25:15.223537000 +0800
+++ new/mt7996/mcu.c	2025-06-23 15:37:29.328485000 +0800
@@ -586,6 +586,9 @@
 		if (gi)
 			tmp_rate.flags |= RATE_INFO_FLAGS_SHORT_GI;
 		break;
+		
+	case MT_PHY_TYPE_PLR:
+		break;
 	case MT_PHY_TYPE_HE_SU:
 	case MT_PHY_TYPE_HE_EXT_SU:
 	case MT_PHY_TYPE_HE_TB:
