diff -uNr old/mt7996/mac.c new/mt7996/mac.c
--- old/mt7996/mac.c	2025-06-11 11:08:34.099069000 +0800
+++ new/mt7996/mac.c	2025-06-12 20:55:37.823704000 +0800
@@ -1437,15 +1437,16 @@
 		goto unlock;
 	}
 
-#ifdef CONFIG_NET_MEDIATEK_SOC_WED	
-	if (mtk_wed_device_active(&dev->mt76.mmio.wed) && wcid->sta) {
+//#ifdef CONFIG_NET_MEDIATEK_SOC_WED	
+	//if (mtk_wed_device_active(&dev->mt76.mmio.wed) && wcid->sta) {
+	if (wcid->sta) {	
 		/* Do not check TX BA status for mgmt frames which are sent at a
 		 * fixed rate
 		 */
 		if (!le32_get_bits(txs_data[3], MT_TXS3_FIXED_RATE))
 			mt7996_check_tx_ba_status(wcid, FIELD_GET(MT_TXS0_TID, txs));
 	}
-#endif
+//#endif
 	txrate = FIELD_GET(MT_TXS0_TX_RATE, txs);
 	bw = FIELD_GET(MT_TXS0_BW, txs);
 
diff -uNr old/mt7996/mt7996.h new/mt7996/mt7996.h
--- old/mt7996/mt7996.h	2025-06-11 11:08:34.048042000 +0800
+++ new/mt7996/mt7996.h	2025-06-12 20:58:08.235686000 +0800
@@ -1415,10 +1415,12 @@
 static inline bool
 mt7996_vow_should_enable(struct mt7996_dev *dev)
 {
-	#ifdef CONFIG_NET_MEDIATEK_SOC_WED	
-	return !wiphy_ext_feature_isset(mt76_hw(dev)->wiphy,
-	                                NL80211_EXT_FEATURE_AIRTIME_FAIRNESS) ||
-	       mtk_wed_device_active(&dev->mt76.mmio.wed);
+	//#ifdef CONFIG_NET_MEDIATEK_SOC_WED
+	#if 1
+	//return !wiphy_ext_feature_isset(mt76_hw(dev)->wiphy,
+	  //                              NL80211_EXT_FEATURE_AIRTIME_FAIRNESS) ||
+	      // mtk_wed_device_active(&dev->mt76.mmio.wed);
+	    return true;
 	#else
 	return !wiphy_ext_feature_isset(mt76_hw(dev)->wiphy,
 	                                NL80211_EXT_FEATURE_AIRTIME_FAIRNESS);
